allennlp/semparse/worlds/atis_world.py

```diff
@@ -31,74 +59,182 @@ def __init__(self,
                  utterances: List[str],
                  tokenizer: Tokenizer = None,
                  database_directory: str = None) -> None:
+
+        self.all_tables = ALL_TABLES
+        self.tables_with_strings = TABLES_WITH_STRINGS
         if database_directory:
-            self.sql_table_context = SqlTableContext(ALL_TABLES,
-                                                     TABLES_WITH_STRINGS,
-                                                     database_directory)
+            self.database_directory = database_directory
+            self.connection = sqlite3.connect(self.database_directory)
+            self.cursor = self.connection.cursor()
+
+        self.sql_table_context = SqlTableContext(ALL_TABLES,
+                                                 TABLES_WITH_STRINGS,
+                                                 database_directory,
+                                                 utterances=utterances) if database_directory else None
+
+        self.grammar_dictionary = deepcopy(self.sql_table_context.grammar_dictionary)
 
         self.utterances: List[str] = utterances
         self.tokenizer = tokenizer if tokenizer else WordTokenizer()
         self.tokenized_utterances = [self.tokenizer.tokenize(utterance) for utterance in self.utterances]
-        valid_actions, linking_scores = self.init_all_valid_actions()
-        self.valid_actions: Dict[str, List[str]] = valid_actions
+        self.linked_entities = self.get_linked_entities()
```

Comment:
This method, and `update_valid_actions` and `flatten_entities`, look like they should be private.